### PR TITLE
Simplified the parameterized loss computation.

### PR DESCRIPTION
**Pull request by ondrasej**
Streamlines the computation of the loss and brings it more in line with an eager
TF2 style. 

As a side effect, I've added some duplication of work in repeated normalization of
the outputs. This could be avoided by unwrapping some of the computation, but it's
probably not worth the extra code and complexity. And, at least in theory, the
duplicate ops should be removed by the JIT graph optimizer.

This is a follow-up to ##343.

### CODE REVIEW

File: gematria/model/python/loss_utils.py [hunk 1]
```diff
@@ -175,7 +167,7 @@ class LossComputation(tf.experimental.ExtensionType):
   absolute_error_percentiles: tf.Tensor
   absolute_percentage_error_percentiles: tf.Tensor
 
-
+@tf.function
```
Comment (boomanaiden154) [inline gematria/model/python/loss_utils.py:170]: I'm not sure a `tf.function` annotation here is useful. Tracing for tests I would guess just increases the time they take to run. This will also get covered by the `tf.function` annotation on the training step when actually running the model.

File: gematria/model/python/loss_utils.py [hunk 2]
```diff
@@ -138,30 +68,92 @@ def _make_percentile_tensor(
         tfp.stats.percentile(task_values, percentile_ranks)
     )
     assert percentile_tensors[-1].shape.is_compatible_with((None,))
-  return tf.stack(percentile_tensors, axis=1)
+  percentile_tensor = tf.stack(percentile_tensors, axis=1)
+  assert not percentile_ranks or percentile_tensor.shape.is_compatible_with(
```
Comment (boomanaiden154) [inline gematria/model/python/loss_utils.py:72]: Is it possible to change this to an `if` statement that returns a `ValueError` if the conditions aren't met? Python `assert`s are kind of a pain with not returning call stacks and what not.

File: gematria/model/python/loss_utils.py [hunk 3]
```diff
@@ -138,30 +68,92 @@ def _make_percentile_tensor(
...
+) -> tf.Tensor:
+  """Applies the selected loss function to normalized absolute deltas."""
+  match loss_type:
+    case options.LossType.MEAN_SQUARED_ERROR:
+      return tf.square(normalized_delta)
+    case options.LossType.MEAN_ABSOLUTE_ERROR:
+      return tf.abs(normalized_delta)
+    case options.LossType.HUBER:
+      return _huber(tf.abs(normalized_delta))
+  raise ValueError('Unexpected loss type {loss_type!r}')
```
Comment (boomanaiden154) [inline gematria/model/python/loss_utils.py:90]: It looks like this is intended to be an f-string?

File: gematria/model/python/loss_utils.py [hunk 4]
```diff
@@ -138,30 +68,92 @@ def _make_percentile_tensor(
...
-):
-  """Returns a tensor of normalized errors."""
+    delta: tf.Tensor,
+    expected_outputs: tf.Tensor,
+) -> tf.Tensor:
+  """Applies the given error normalization function to deltas.
+
+  Args:
+    normalization: The selected normalization function.
+    num_tasks: The number of tasks used in the model.
```
Comment (boomanaiden154) [inline gematria/model/python/loss_utils.py:102]: It looks like this got removed?

**Review by boomanaiden154 — COMMENTED** [review]
Some comments. This definitely looks like a pretty significant improvement.

File: gematria/model/python/loss_utils.py (hunk 2 above)
Comment (ondrasej) [reply gematria/model/python/loss_utils.py:72]: This is weird - a failing `assert` raises an exception (`AssertionError`) that behaves like any other Python exception, so it should have more or less the same effects as raising a `ValueError`.

As for turning this into a `ValueError` - that should be raised when the inputs of a function are invalid, which is not quite the case with this assert (the assert should never fail unless there's a bug in the code).

That said, we could get rid of a lot of the `num_tasks` parameters here, and simplify the code some more. Perhaps let me do that in another cleanup.

File: gematria/model/python/loss_utils.py (hunk 3 above)
Comment (ondrasej) [reply gematria/model/python/loss_utils.py]: Done. Thanks for catching it!

File: gematria/model/python/loss_utils.py (hunk 2 above)
Comment (boomanaiden154) [reply gematria/model/python/loss_utils.py:72]: It looks like my recollection is inaccurate. Assuming it raises an exception, this looks good to me.

**Review by boomanaiden154 — APPROVED** [review]
